pkg/apis/pipeline/v1alpha1/taskrun_validation_test.go

```diff
@@ -0,0 +1,326 @@
+package v1alpha1
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/knative/pkg/apis"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestTaskRunValidate(t *testing.T) {
+	tests := []struct {
+		name string
+		task TaskRun
+		want *apis.FieldError
+	}{
+		{
+			name: "invalid taskspec",
+			task: TaskRun{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "taskmetaname",
+				},
+			},
+			want: apis.ErrMissingField("spec"),
+		},
+		{
+			name: "invalid taskrun metadata",
+			task: TaskRun{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "task.name",
+				},
+			},
+			want: &apis.FieldError{
+				Message: "Invalid resource name: special character . must not be present",
+				Paths:   []string{"metadata.name"},
+			},
+		},
+	}
+
+	for _, ts := range tests {
+		t.Run(ts.name, func(t *testing.T) {
+			err := ts.task.Validate()
+			if d := cmp.Diff(err.Error(), ts.want.Error()); d != "" {
+				t.Errorf("Validate/%s (-want, +got) = %v", ts.name, d)
+			}
+		})
+	}
+}
+
+func TestTaskRunSpecValidate(t *testing.T) {
+	tests := []struct {
+		name    string
+		spec    *TaskRunSpec
+		wantErr *apis.FieldError
+	}{
+		{
+			name:    "invalid taskspec",
+			spec:    &TaskRunSpec{},
+			wantErr: apis.ErrMissingField("spec"),
+		},
+		{
+			name: "invalid taskref",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{},
+				Trigger: TaskTrigger{
+					TriggerRef: TaskTriggerRef{
+						Type: TaskTriggerTypePipelineRun,
+					},
+				},
+			},
+			wantErr: apis.ErrMissingField("spec.taskref.name"),
+		},
+		{
+			name: "invalid taskref",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{
+					Name: "taskrefname",
+				},
+				Trigger: TaskTrigger{
+					TriggerRef: TaskTriggerRef{
+						Type: "wrongtype",
+					},
+				},
+			},
+			wantErr: apis.ErrInvalidValue("wrongtype", "spec.trigger.triggerref.type"),
+		},
+		{
+			name: "valid task trigger run type",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{
+					Name: "taskrefname",
+				},
+				Trigger: TaskTrigger{
+					TriggerRef: TaskTriggerRef{
+						Type: TaskTriggerTypePipelineRun,
```

Comment:
if the type is `pipelinerun` there should be a `name` field as well with the name of the pipelinerun